tokio/tests/rt_threaded.rs

```diff
@@ -692,6 +692,77 @@ fn mutex_in_block_in_place() {
     })
 }
 
+// Tests that when a task is notified by another task and is placed in the LIFO
+// slot, and then the notifying task blocks the runtime, the notified task will
+// be stolen by another worker thread.
+//
+// Integration test for: https://github.com/tokio-rs/tokio/issues/4941
+#[test]
+fn lifo_stealable() {
+    use std::time::Duration;
+
+    let (unblock_tx, unblock_rx) = tokio::sync::oneshot::channel();
+    let (task_started_tx, task_started_rx) = tokio::sync::oneshot::channel();
+    let (block_thread_tx, block_thread_rx) = mpsc::channel::<()>();
+    let rt = runtime::Builder::new_multi_thread()
+        // Make sure there are enough workers that one can be parked running the
+        // I/O driver and another can be parked running the timer wheel and
+        // there's still at least one worker free to steal the blocked task.
+        .worker_threads(4)
+        .enable_time()
+        .build()
+        .unwrap();
+
+    rt.block_on(async {
+        // Keep the runtime busy so that the workers that might steal the
+        // blocked task don't all park themselves forever.
```

Comment:
```suggestion
        // Keep the runtime busy so that the workers that might steal the
        // blocked task don't all park themselves forever.
        //
        // Since multi-thread scheduler doesn't push to the LIFO slot if it
        // it not empty, so this task will be pushed into the global inject
        // queue eventually, and unpark another worker thread that will steal
        // the `blocked_task_joined` from the LIFO slot.
```

I added some comments to help me to understand this test, but it looks too long, feel free to rewrite it.

Reply:
As I understand it, the reason this task doesn't go in the LIFO slot isn't because of the injector queue behavior, it's because it's being woken up by a timer,  and is therefore woken by whatever worker thread ends up running the time driver, rather than the worker that was last running this task.